### ISSUE TITLE
[Bug]: useLocation inside <Routes location={location}> doesn't return given location

### ISSUE THREAD

**Issue by gitcatrat** [remix-run/react-router#8470]
### What version of React Router are you using?

6.1.1

### Steps to Reproduce

Full reproduction can be found here, it's functionally similar to v5 modal gallery example:
https://codesandbox.io/s/heuristic-rhodes-508k5?file=/src/App.js

Steps to replicate:
1. Click on **_Post_** link
2. Home page should now use previous "background" location
3. Notice how `useLocation` in Home still returns new location 

### Expected Behavior

`useLocation` inside `<Routes location={location}>` returns the `prop` location.
`useLocation` inside `<Routes>` returns the "correct" currently active location.

### Actual Behavior

Both `useLocation` return "correct" currently active location (as the one corresponding to URL bar).

I.e issue seems to be that `useLocation` is not scoped, it only reads global location.

**Comment by gitcatrat** [comment]
`useLocation` is reading straight from context and I don't see anything updating context `location`, so I'm guessing that it has never returned "scoped" location in v6. `location` prop was added to `Routes` in late alpha, it's quite understandable that this was overlooked (presuming you agree that this is a bug).

```js
export function useLocation(): Location {
  invariant(
    useInRouterContext(),
    // TODO: This error is probably because they somehow have 2 versions of the
    // router loaded. We can help them understand how to avoid that.
    `useLocation() may be used only in the context of a <Router> component.`
  );

  return React.useContext(LocationContext).location;
}
````

**Comment by liuhanqu** [comment]
```ts
  let location = React.useMemo(() => {
    let trailingPathname = stripBasename(pathname, basename);

    if (trailingPathname == null) {
      return null;
    }

    return {
      pathname: trailingPathname,
      search,
      hash,
      state,
      key
    };
  }, [basename, pathname, search, hash, state, key]);
```

The `location` would change if it's dependencies change. You can see the code is in the `Router` component.

**Comment by gitcatrat** [comment]
@liuhanqu Yup you're right. But the issue here seems to be that there's only one location in context. 

I expect this behavior:
`useLocation` inside `<Routes location={location}>` returns the prop location.
`useLocation` inside `<Routes>` returns the "correct" currently active location.

**Comment by gitcatrat** [comment]
I had to use a temporary workaround for now because this is blocking my huge release!

1. wrap the `Routes` with `location` prop with custom `Context` and pass the same prop into it.
```js
export const Location = createContext();

function Routes() {
  const location = useLocation();
  let background = null;
  
  if (location?.state?.background) {
    // "layered" logic might work differently in your app but
    // when modal is opened, I've got the following state:
    // {modal: true, background: previousLocation}
    background = location.state.background;
  }

  return (
    <Location.Provider value={background || location}> {/* this right here officer */}
     <Routes location={background || location}>  
      {/* normal/background routes */}
     </Routes> 
   </Location.Provider>
   
   {background &&
     <Routes>
      {/* overlay/modal routes */}
     </Routes>
   }
 );
}
```

2. Make a custom `useLocation` that tries to figure out where it is based on context existence.
```js
export default function useLocation() {
  const scopedLocation = useContext(Location);
  const globalLocation = useReactRouterLocation(); // "real" useLocation 
  const modalIsOpen = scopedLocation?.pathname !== globalLocation?.pathname;
  const thisHookIsCalledFromModal = modalIsOpen && !scopedLocation && globalLocation?.state?.modal;
  return (thisHookIsCalledFromModal ? globalLocation : scopedLocation);
}
```

Dadaa! Now `useLocation` returns the correct `location` in every situation!

**Comment by gitcatrat** [comment]
I had to use a temporary workaround for now because this is blocking my huge release!

1. wrap the `Routes` with `location` prop with custom `Context` and pass the same prop into it.
```js
export const Location = createContext();

function Routes() {
  const location = useLocation();
  let background = null;
  
  if (location?.state?.background) {
    // "layered" logic might work differently in your app but
    // when modal is opened, I've got the following state:
    // {modal: true, background: previousLocation}
    background = location.state.background;
  }

return (
  <>
    <Location.Provider value={background || location}> {/* this right here officer */}
      <Routes location={background || location}>  
        {/* normal/background routes */}
      </Routes> 
    </Location.Provider>
  
    {background &&
      <Routes>
        {/* overlay/modal routes */}
      </Routes>
    }
  </>
);
}
```

2. Make a custom `useLocation` that tries to figure out where it is based on context existence.
```js
export default function useLocation() {
  const scopedLocation = useContext(Location);
  const globalLocation = useReactRouterLocation(); // "real" useLocation 
  const modalIsOpen = scopedLocation?.pathname !== globalLocation?.pathname;
  const thisHookIsCalledFromModal = modalIsOpen && !scopedLocation && globalLocation?.state?.modal;
  return (thisHookIsCalledFromModal ? globalLocation : scopedLocation);
}
```

Dadaa! Now `useLocation` returns the correct `location` in every situation!

**Comment by bloomdido** [comment]
Gah! This bug just cost me a few hours...

Thanks @gitcatrat for reporting and giving a workaround!

**Comment by bloomdido** [comment]
Unfortunately, I'm using `useSearchParams()` to read values out of the `location`, which is coded to use the unscoped `useLocation()` hook, so I'm going to have to do a bit more work on my end.

**Comment by mrwwalmsley** [comment]
This issue also cost me a few hours!

I'm also using `useSearchParams()` to read values out of the `location`.  I am using a CSS  TransitionBlock

```
 <TransitionBlock id={location.key || ' '} fullScreen>
      <Routes location={location}>
        <Route path="/menu/a" element={<A />} />
        <Route path="/menu/b" element={<B />} />
      </Routes>
    </TransitionBlock>
```

While navigating from page A to page B, both pages render with the current `location` state.
This was not the behaviour with the `Switch`.

This is the last thing I have to do to complete upgrade to ReactRouter v6.

**Comment by SleeplessByte** [comment]
In the source code we can see that indeed `useLocation` does _not_ use the injected `location` prop. Without creating the custom hook like @gitcatrat described, you can do this:

```typescript
import { Action } from 'history';
import {  UNSAFE_LocationContext as LocationContext } from 'react-router-dom'

// wherever you need the stored location, or anywhere higher in the tree
<LocationContext.Provider value={{ location: storedLocation, navigationType: Action.Pop }}>
  {...}
</LocationContext.Provider>
```

I expected #8008 to mention this issue as well.

**Comment by evoyy** [comment]
Well, this sucks. This breaks functionality for anyone using the background modal technique that dates all the way back to React Router v1. The modal example in the official documentation is also affected by this.

As mentioned above, the location object given as a prop to Routes and then passed on to useRoutes, is not used. Only the pathname property of the passed location is used.

The workaround using a custom location context feels a bit hackish and requires making changes to any components that use location for logic (or undesired things will happen behind the modal). I considered reverting to v5 but for now I have simply reset my isModal state to false before rendering my routes, which causes a navigation instead of opening the modal.

Michael or Tim @mjackson @timdorr, please could you share your opinion on this issue?

**Comment by SleeplessByte** [comment]
You don't need to make changes to all components that use location for logic. You only need it around the modals. A single context provider would work.

(That said, it feels very hacky)

**Comment by evoyy** [comment]
@SleeplessByte I tried your workaround but it doesn't seem to work with components that use other hooks such as useParams(). The path matching seems to be broken. It works when I create my own LocationContext and use it to pass down my stored location:

````javascript
export const LocationContext = createContext();
````

````javascript
<LocationContext.Provider value={isModal ? prevLocation : location}>
...
</LocationContext.Provider>
````

Then I have to do this across my whole codebase:

````diff
-  import { useLocation } from 'react-router-dom';
+  import { LocationContext } from 'myApp';

-  const location = useLocation();
+  const location = useContext(LocationContext);
````

**Comment by SleeplessByte** [comment]
I think the provider may not be at the "right" location then. I use this workaround in production.

Using your own context (and hook) definitely should work 😁

**Comment by evoyy** [comment]
@SleeplessByte You were right, the scope of my provider was too large and so I was passing the wrong location to my modal (God, this is confusing!!). Your workaround is a nice solution. Thanks for your help!

The only change, instead of importing Action from history, I'm getting the navigation type from the hook:

````javascript
const navigationType = useNavigationType();
````

**Comment by SleeplessByte** [comment]
Yeah. That should work. Instead of hardcoding it, that should actually retrieve the value dynamically.

**Comment by loicmagne** [comment]
any plans to solve this ?

**Comment by zhouzi** [comment]
It would be great to know if the maintainers agree that it's a bug and that `<Routes />` should provide the new location through the location context. We could then open a PR to propose a fix (I would be happy to give it a try).

**Comment by brophdawg11** [comment]
I'm going to defer to @mjackson and @ryanflorence here (as they've got the historical insight) but this doesn't feel like a bug to me.  The `Location` concept is basically a wrapper around `window.location` which is a singleton -- it's impossible to have 2 separate `window.location`'s and thus it feels odd to me if `useLocation` can actually have different contextual values based on where you are in the tree.

If we made the change in #9094, wouldn't that just introduce the opposite problem in that contextual components could _never_ actually access the true global location anymore?  Seems like we're trading one problem for another.

In the example provided, couldn't components that need to use a contextual location just do something like the following:

```js
const location = useLocation();
const contextualLocation = location.state?.background || location; 
```

That now leaves the decision of which `location` to use in the hands of the developer without prohibiting contextual components from accessing the global location?

**Comment by SleeplessByte** [comment]
It feels very weird to have to:

- either import `UNSAFE_XXX`...
- or create a new wrapper and new useXXX hooks.

I would be completely okay with keeping the workaround I listed, but I do think something needs to be done in order to support https://github.com/remix-run/react-router/discussions/8008.

**Comment by zhouzi** [comment]
> it's impossible to have 2 separate `window.location`'s and thus it feels odd to me if `useLocation` can actually have different contextual values based on where you are in the tree.

This is particularly useful for when you have a modal that has its own URL. For example: https://v5.reactrouter.com/web/example/modal-gallery

> If we made the change in #9094, wouldn't that just introduce the opposite problem in that contextual components could _never_ actually access the true global location anymore?

The idea is that the components that are receiving the "background" location work as usual. They are not aware of the concept of background/foreground location. To be honest I have never come across a use case where the background location would need to access the "global" location.

> ```js
> const location = useLocation();
> const contextualLocation = location.state?.background || location; 
> ```
> 
> That now leaves the decision of which `location` to use in the hands of the developer without prohibiting contextual components from accessing the global location?

The problem is that the components where we need this behavior are `<Route />` and anything related to the location (e.g hooks). For example:

```
<Routes location={location.state?.background ?? location}>
  <Route path="/first" element={<Page1 />} exact /> // Route should read from the location above
</Routes>
```

**Comment by johnpangalos** [comment]
@brophdawg11 I can change #9094 to add the global location to the LocationContext and add a hook to fetch global location if someone would want that. That way the current behavior could still be accessed if needed.

**Comment by flying-sheep** [comment]
I totally agree with @zhouzi and @johnpangalos.

@brophdawg11 my intuition about `Location` is the same as [`react-helmet-async`](https://www.npmjs.com/package/react-helmet-async), i.e. there’s a “global default” derived from browser APIs. (`window.location` here, `window.document`’s DOM for Helmet). Users (or library authors) can choose to override those values. The truth is whatever is set in the closest parent (in case of multiple nested ones) and if there’s no override, the browser’s global default wins out.

I didn’t consider the possiblility that the current behavior might possibly be intentional at all. I went in here with a very clear feeling of “ah, good thing this bug is finally being addressed”. As you can see from @zhouzi’s comment, merging #9094 would allow library authors to rely on the ability to truly override the hook’s return value. This alone is an indicator of good design.

**Comment by mycroes** [comment]
> I'm going to defer to @mjackson and @ryanflorence here (as they've got the historical insight) but this doesn't feel like a bug to me. The `Location` concept is basically a wrapper around `window.location` which is a singleton -- it's impossible to have 2 separate `window.location`'s and thus it feels odd to me if `useLocation` can actually have different contextual values based on where you are in the tree.

The gallery example uses a different set of routes for the modal and for the background, I guess the fact that you can use different routes within the same top-level router alone warrants `useLocation` to be contextual. I've been using similar approaches with 2 `Routes` (or `Switch` in v5), one for rendering page titles and the other for rendering page content. In all honesty I don't have `useLocation` issues in that specific situation, but I guess it makes sense to have `useLocation` aware of the `Routes` it's in.

**Comment by AsemK** [comment]
This issue just caused me several hours of debugging. It is a major breaking change that is never mentioned in the documentation. I agree with others that the most intuitive behavior is that `useLocation` and `useSearchParams` should return values based on the context not on the global location object.
For now whenever I use `useLocation`, I have to do this:
```js
const globalLocation = useLocation();
const location = globalLocation.state?.backgroundLocation || globalLocation;
```
Any plans for fixing that? I think most use cases of `useLocation` would expect the contextual location.

The interface mentions:
```js
/**
 * Returns the current location object, which represents the current URL in web
 * browsers.
 *
 * Note: If you're using this it may mean you're doing some of your own
 * "routing" in your app, and we'd like to know what your use case is. We may
 * be able to provide something higher-level to better suit your needs.
 *
 * @see https://reactrouter.com/docs/en/v6/api#uselocation
 */
 export declare function useLocation(): Location;
```
I would argue that, in order to do "some of your own routing", you need the contextual location.

**Comment by evoyy** [comment]
@AsemK The same thing happened to me. I believe this will be resolved once the following pull request is merged: https://github.com/remix-run/react-router/pull/9094

**Comment by brophdawg11** [comment]
Hey folks - I checked in with @ryanflorence and he confirmed this is a bug, so we should be able to get #9094 merged shortly for release with `6.4.0` 👍
